projects/packages/forms/src/contact-form/class-contact-form-plugin.php

```diff
@@ -749,46 +749,300 @@ public static function gutenblock_render_form_step_navigation( $atts, $content )
+			$classes       = $processor->get_attribute( 'class' );
+			$is_dots_style = strpos( $classes, 'is-style-dots' ) !== false;
+		}
+
+		// Reset processor
 		$processor = new \WP_HTML_Tag_Processor( $content );
 		$processor->next_tag();
 		$processor->set_attribute( 'data-wp-interactive', 'jetpack/form' );
 
+		// Process legacy progress bar elements
 		while ( $processor->next_tag() ) {
 			$class = $processor->get_attribute( 'class' );
 			if ( 'jetpack-form-progress-indicator-bar' === $class ) {
 				$processor->set_attribute( 'data-wp-style--width', 'state.getStepProgress' );
 			}
 		}
 
-		return $processor->get_updated_html();
+		// Get the updated HTML after processing attributes
+		$updated_content = $processor->get_updated_html();
+
+		// Build steps HTML using the actual number of form steps
+		$steps_html = '';
+		for ( $i = 0; $i < $max_steps; $i++ ) {
+			$step_classes = 'jetpack-form-progress-indicator-step';
+			if ( $i === 0 ) { // First step is active by default
+				$step_classes .= ' is-active';
+			}
+
+			$step_context = array( 'stepIndex' => $i );
+
+			$steps_html .= sprintf(
+				'<div class="%s" data-wp-class--is-active="state.isStepActive" data-wp-class--is-completed="state.isStepCompleted" data-wp-context=\'%s\' data-step-index="%d">',
+				esc_attr( $step_classes ),
+				wp_json_encode( $step_context ),
+				$i
+			);
+
+			// Add line element for all styles
+			$steps_html .= '<div class="jetpack-form-progress-indicator-line"></div>';
+
+			// Add dot element only for dots style
+			if ( $is_dots_style ) {
+				$steps_html .= '<div class="jetpack-form-progress-indicator-dot">';
+				$steps_html .= sprintf(
+					'<span class="jetpack-form-progress-indicator-step-number" data-wp-text="state.getStepContent" aria-label="%s">%d</span>',
+					// Translators: %d: The step number in the progress indicator.
+					esc_attr( sprintf( __( 'Step %d', 'jetpack-forms' ), $i + 1 ) ),
+					$i + 1
+				);
+				$steps_html .= '</div>';
+			}
+
+			$steps_html .= '</div>';
+		}
+
+		// Add progress element for both styles (uses different state getters)
+		$progress_state   = $is_dots_style ? 'state.getDotsProgress' : 'state.getStepProgress';
+		$progress_element = '<div class="jetpack-form-progress-indicator-progress" data-wp-style--width="' . $progress_state . '"></div>';
+
+		// Replace the empty steps container with the populated one
+		$updated_content = str_replace(
+			'<div class="jetpack-form-progress-indicator-steps"></div>',
+			'<div class="jetpack-form-progress-indicator-steps">' . $progress_element . $steps_html . '</div>',
+			$updated_content
+		);
+
+		return $updated_content;
+	}
+
+	/**
+	 * Find the parent contact form block for a given progress indicator block.
+	 *
+	 * @param WP_Block $block The progress indicator block.
+	 * @return array|null The parent form block or null if not found.
+	 */
+	private static function find_parent_form_block( $block ) {
```

Comment:
I’m thinking about doing this differently. Maybe we store the number of steps as an attribute instead of including all this code to calculate it. That said, the current approach might still be useful for the follow-up that adds step names. For that, though, we could probably rely on frontend js to fetch the step labels.